File: 프로그래머스/2/154539. 뒤에 있는 큰 수 찾기/뒤에 있는 큰 수 찾기.py
def solution(numbers):
    n = len(numbers)
    answer = [-1] * n
    stack = []
    
    for i in range(n):
        while stack and numbers[i] > numbers[stack[-1]]:
            answer[stack[-1]] = numbers[i]
            stack.pop()
        stack.append(i)
        
    return answer

File: 프로그래머스/2/154539. 뒤에 있는 큰 수 찾기/뒤에 있는 큰 수 찾기.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <stack>
#include <iostream>

using namespace std;

vector<int> solution(vector<int> numbers) {
    stack<int> st;
    int n = numbers.size();
    vector<int> answer(n, -1);
    
    for(int i = 0; i < n; i++){
        while(!st.empty() && numbers[i] > numbers[st.top()]){
            answer[st.top()] = numbers[i];
            st.pop();
        }
        st.push(i);
    }
    return answer;
}